### ISSUE TITLE
Completely flawed benchmark, nothing about the parser is stressed

### ISSUE THREAD

**Issue by ghost** [felixge/faster-than-c#7]
I consider myself a (skeptic) performance enthusiast and I wonder what happened to your latest parser? I want to know because this whole talk kind of relies on you actually having a real project to show or not. It seems you still maintain and develop the node-mysql project, did this project ever end up being faster than the native mariasql?

The last graph shows good performance but you also mentioned that you had memory issues. How did this whole "faster-than-C" turn out in reality? Did you manage to write something faster than the native mariasql that worked out as a real project?

**Comment by felixge** [comment]
> I consider myself a (skeptic) performance enthusiast and I wonder what happened to your latest parser?

I stopped working with node.js a few months after working on this, so the latest parser was never integrated into the main project.

> I want to know because this whole talk kind of relies on you actually having a real project to show or not.

It doesn't rely depend on that. The new parser didn't implement all parts of the protocol, but those wouldn't have had a performance impact for parsing row data.

> It seems you still maintain and develop the node-mysql project, did this project ever end up being faster than the native mariasql?

I haven't maintained the project since late 2012. Community members have continued to maintain the project.

> The last graph shows good performance but you also mentioned that you had memory issues.

My benchmarks uncovered what looked like v8 GC problems. I discussed them with Erik Corry in person at the time, but I'm not sure what came out of it.

> How did this whole "faster-than-C" turn out in reality? Did you manage to write something faster than the native mariasql that worked out as a real project?

This repository contains the latest results from this project.

**Comment by ghost** [comment]
So basically `mariasql` is still the fastest library and not any of the JavaScript based libraries you worked on. That sentence alone kind of speaks for iteself in my opinion. Blaming on the "V8 GC" is not really a valid excuse for using tons of memory - you know that there exists _far_ faster memory allocation algorithms for C to use if one does not care that much about memory usage (pool allocators waste more memory but are _a lot_ faster).

~~I don't see any code or any benchmark code so I cannot validate your results and I cannot reimplement your parser in C, so that is also something that seems kind of fishy.~~

You also do compare both solutions using _JavaScript_ end result. What I mean is, both products are JavaScript bottlenecked by having to create JavaScript obects and not native objects for each language. I know from experience that as soon as your native code touches JavaScript in any way (calling JS callbacks, creating JS objects, etc), you lose at least 5x performance and 5x memory usage. A more sane comparison would be "how fast can a _native_ C solution be, compared to a JS solution", instead of "how fast can C create _JavaScript objects_, compared to how fast JavaScript can create _JavaScript objects_". You see my point here?

Edit: Holy shit, you actually seem to have some code here. I completely missed that. So, how would I run this?

**Comment by ghost** [comment]
I'm trying to run your code but it fails like this:

```
[alexhultman@localhost benchmark]$ node run.js new-parser
/home/alexhultman/faster-than-c/figures/mysql2-vs-new-parser/benchmark/new-parser/ElementParser.js:129
  var value = this._buffer.parent.utf8Slice(offset, end);
                                 ^

TypeError: Cannot read property 'utf8Slice' of undefined
    at ElementParser.parseString (/home/alexhultman/faster-than-c/figures/mysql2-vs-new-parser/benchmark/new-parser/ElementParser.js:129:34)
    at ElementParser.parseLengthCodedString (/home/alexhultman/faster-than-c/figures/mysql2-vs-new-parser/benchmark/new-parser/ElementParser.js:86:15)
    at Protocol.eval [as _parseRow] (eval at <anonymous> (/home/alexhultman/faster-than-c/figures/mysql2-vs-new-parser/benchmark/new-parser/Protocol.js:90:20), <anonymous>:3:21)
    at Protocol.parseRow (/home/alexhultman/faster-than-c/figures/mysql2-vs-new-parser/benchmark/new-parser/Protocol.js:58:19)
    at Protocol.write (/home/alexhultman/faster-than-c/figures/mysql2-vs-new-parser/benchmark/new-parser/Protocol.js:30:32)
    at FixtureStream.<anonymous> (/home/alexhultman/faster-than-c/figures/mysql2-vs-new-parser/benchmark/new-parser/index.js:16:16)
    at emitOne (events.js:90:13)
    at FixtureStream.emit (events.js:182:7)
    at FixtureStream.resume (/home/alexhultman/faster-than-c/figures/mysql2-vs-new-parser/benchmark/FixtureStream.js:31:10)
    at /home/alexhultman/faster-than-c/figures/mysql2-vs-new-parser/benchmark/run.js:25:10
```

**Comment by ghost** [comment]
It seems `this._buffer.parent` in `ElementParser.parseString` is undefined.

Edit: I fixed it by replacing this:

``` javascript
  //var value = this._buffer.parent.utf8Slice(offset, end);
  var value = this._buffer.toString('utf8', offset, end);
```

Do you think that will change your perf? It runs now at least.

**Comment by ghost** [comment]
new-parser runs quite steadily at around 200 mb of memory, it does increase slightly over time and it makes use of 100% CPU. The whole run takes 1 minute 30 seconds for me - does that seem reasonable, or did I mess it up somehow? I ran it in Node.js 5.11.

How do I check the output for correctness? Shoud it produce some kind of output that I can easily "diff" with my port? What happens if I come up with a much faster parser? Will you update your post then?

**Comment by felixge** [comment]
@alexhultman unfortunately I don't have time to help getting this code updated to with the node.js changes of the last 4 years. Your change may impact performance.

That being said, you're missing the point of this project. Comparing JS vs Native C performance was never the goal. The project's goal was to determine if a pure (Node)JS MySQL client could outperform a Node.js MySQL Client that calls out to an existing C library. My conclusion in 2012 was, that yes, it's doable. As you pointed out, this can be partially explained by the C version still needing to convert its end results into JS objects, paying similar, or even higher, overhead costs in the process.

> new-parser runs quite steadily at around 200 mb of memory, it does increase slightly over time and it makes use of 100% CPU. The whole run takes 1 minute 30 seconds for me - does that seem reasonable, or did I mess it up somehow? I ran it in Node.js 5.11.

Seems reasonable. How many rows/sec?

> How do I check the output for correctness? Shoud it produce some kind of output that I can easily "diff" with my port? What happens if I come up with a much faster parser? Will you update your post then?

I don't remember how I verified the correctness. IIRC the parser code itself had separate tests. I may have verified the benchmark results manually.

**Comment by ghost** [comment]
I get that, I have read your post like 3 times. It clearly states it is kind of a click-bait title (which I agree on). But I still want to 1) outperform it, 2) maybe even outperform it as an addon. I looked at the code and the whole parser is incredibly small in code size so it seems like a very reasonable task.

You clearly utilize eval as some kind of explicit JIT:ing, which I'm familiar with from OpenCL (I've written code that recompiles based on the input, which have outperformed a statically compiled binary). I plan on doing the same here if it's needed - one can use a C compiler to JIT C code that is optimized for the specific input.

So okay, each row in the output is a complete benchmark and you run multiple runs. It looks like this, I don't know what it means (excerpt):

```
benchmark   number  duration    time    rss heapUsed    heapTotal   nodeVersion v8Version   bytes   rows
new-parser  1   129 1471342694378   208855040   5416584 10619424    5.11.0  4.6.85.31180821863  100000
new-parser  2   117 1471342694502   214605824   8336048 14747168    5.11.0  4.6.85.31180821863  100000
new-parser  3   120 1471342694622   215416832   8562024 14747168    5.11.0  4.6.85.31180821863  100000
new-parser  4   117 1471342694740   215416832   8406208 14747168    5.11.0  4.6.85.31180821863  100000
```

**Comment by ghost** [comment]
Hey, I just finished my first initial commit: https://github.com/alexhultman/faster-than-faster-than-c

It currently does the same as new-parser in the same "hard-coded" way. It currently parses 10x faster than "new-parser". I should see if I can make it an addon and see how I can expose a good interface in JS-land without losing way too much performance. Anyways, my point is kind of shown here - you can easily get 10x faster in C (well, actually it's C++ but who cares). I haven't fully optimized the parser though.

**Comment by ghost** [comment]
I've cleaned up my repo a bit and looked at exposing the fields to JS in a zero-copy way via Proxies. I still need to implement the streaming interface of Protocol, but I know how this should be done.

If I keep polishing this project up, would you add it to the timeline? A 10x faster Node.js addon that implements the same interface as "new-parser" and emits the same data seems pretty much equally worthy of being in the timeline as "new-parser".

**Comment by ghost** [comment]
As a point of reference, I got 540k rows/sec with "new-parser" and 6.5 million rows/sec with my thing. So it's currently over 12x difference.

**Comment by felixge** [comment]
I had a look at your code. A few things I noticed:
- You seem to pass the entire fixture into your parser [all at once](https://github.com/alexhultman/faster-than-faster-than-c/blob/master/main.cpp#L24). My benchmark uses a chunk size of [64kB](https://github.com/felixge/faster-than-c/blob/master/figures/mysql2-vs-new-parser/benchmark/FixtureStream.js#L8). I assume this is because your parser can't actually handle arbitrary chunk sizes.
- My benchmark has the parser build a JS object [for each row](https://github.com/felixge/faster-than-c/blob/master/figures/mysql2-vs-new-parser/benchmark/new-parser/Protocol.js#L78-L91). Maybe I'm missing it, but I don't see your code building some sort of hash map.
- My parser handles UTF-8 parsing/validation. Yours seems to treat the row values as a raw byte sequence.

Your whole attitude is also starting to rub me the wrong way. What are you trying to accomplish? If you want to challenge my results from 2012, please write a node.js addon (not standalone C parser) against the node.js/v8 version that is indicated in my benchmark results from 2012. If you want to to show that can you write an addon in 2016 that wins against my code from 2012, go ahead. I'll come back in 2020 to outperform your code ;).

Either way, if you come up with an addon that is actually doing a comparable amount of data processing, I'm happy to link to it.

But please consider adopting a less arrogant attitude towards JS. I'm the last to argue that it's a great language, but if you think it's humanities greatest mistake, you ought to open up a history book or something ...

**Comment by ghost** [comment]
It's a one day job, so naturally its not feature complete. I know you pass chunks and that is not implemented yet, like I said. I tried passing one single chunk to your parser and that did not improve perf.

Creating JS objects is not part of the parsers job, like I said I'm looking to create JS proxies instead (zero-copy). Basically you implement operator overloading and just call the provided get function.

You don't do any utf8 validation in the code I ran, that just slices the buffer and even if I completely remove any slicing or string retrieval its still significantly slower.

**Comment by ghost** [comment]
You might also want to remove all that Math.pow kind of madness. What you're doing is basically reading various sizes of unsigned integers, and that's basically all it does. You don't need to implement bit shift in user space code, that's already native to the hardware and not needed if you read integers straight up.

**Comment by ghost** [comment]
"My parser handles UTF-8 parsing/validation. Yours seems to treat the row values as a raw byte sequence."

This is acutally not true. I checked the sources of Node.js 0.8 and followed your utf8Slice down to v8::String::New, which does not check the utf-8 string, it simply swallows whatever bytes you pass it. I tested this in newer V8 and I passed it invalid utf-8 and it didn't change anything. This can also be seen in the interface of the function since it returns a Local<String> and not a Maybe<String>. Also, the documentation simply states "Allocates a new string from UTF-8 data.", so basically you are doing a copy on each string, while the stuff I do does not. Any validation on the data would blow the performance, as I just tested that alone. Validating utf-8 on each string performs far far worse than both of our solutions.

Utf-8 is nothing more than "raw bytes", there's nothing to it. It's nothing more than a chunk of memory interpreted as utf-8. That's why I just traverse the dump and mark where data begins and where it ends. Then this can be accessed from a function call with row and field id. Basically, the mysql dump is designed to allow this, you don't need to perform any copies as the data already exists in memory in a very packed and nice way.

"My benchmark has the parser build a JS object for each row. Maybe I'm missing it, but I don't see your code building some sort of hash map."

You kind of get that for free since you already insist on creating V8 strings as copies of the actual data, instead of changing the interface to allow a zero-copy access. So yeah, I cannot make my parser create JavaScript objects and maintain the performance. Creating JavaScript objects is by far the _only_ bottleneck involved here. Nothing about the parser matters at all if you count the time it takes to shove data into V8. Building that massive JS object with copies of the data as strings is alone 99% of all time spent, so this benchmark is not actually measuring anything about parsing, but rather just measuring how fast you can shove data into V8. The point is to skip shoving data into anything, and just have it accessed from the very source, only then will you be able to stress the _parser_ and not simple data shoving.

This whole talk can basically be summarized as "JavaScript objects are extremely heavy weight and should not be used as a means of handling large quantities of data".

**Comment by felixge** [comment]
> Building that massive JS object with copies of the data as strings is alone 99% of all time spent

I don't think you know how percentages work. This number is completely made up.

That being said, your biggest problem is that you continue to bring up straw man arguments. I'm sorry if node.js is an existential thread to you. I'm not a fan of it either. But this presentation was assuming you're already using node.js, and that you need to access your data as JS objects. So even if I could get over your patronizing attitude, I'm afraid I don't have the time for the shadow boxing.

**Comment by ghost** [comment]
Just a quick question then. Why did you decide to use JavaScript objects as a means of transporting the data into the script?

Because we both know that you could easily get 5 - 10x better perf had you chosen any other interface. Basically you disqualify yourself from any opportunity of getting good perf with that decision.

**Comment by c4milo** [comment]
@alexhultman, seriously, stop being so obnoxious. @felixge's approaches where the best at the time and may not be the best right now. I'm sure there are plenty of undocumented reasons why he used those techniques at the time, that might not be relevant today. But that doesn't give you any right to harass him. Nobody is holding you from creating a new project with better results. 

If you are legitimately curious, please change the tone and we will all learn something useful from this discussion.

**Comment by ghost** [comment]
> @felixge's approaches where the best at the time and may not be the best right now.

Reading an unsigned integer from a Node.js buffer with a loop like this:
1. read 1 byte
2. Math.pow(256, bytesRead) it
3. add result to temporary variable
4. repeat until read the bytes

Is, were and never will be "the best approach at the time". Like I already said, buffer.readUInt16LE existed back in 2011, 1 year prior to this parser being made.

Now, why am I so upset about this little thing? Well, it turns out that the "parser" part of this problem actually consist entirely of reading unsigned integers from the buffer. That is pretty much all there is to the "parsing" of the mysql dump.

Like I demonstrated in my repo, this entire parsing can be done in 15 milliseconds, leaving abouot 200 milliseconds for the rest, which only consist of copying the data into a JavaScript object. That being said, you can clearly determine the parsing to be insignificant in this.

Any benchmark that concludes a solution based on Math.pow to read unsigned integers to be "pretty damn fast" is clearly completely flawed. The only thing being benchmarked here is "how fast can V8 copy strings, and put them in a JS object".

Nothing about this repo benchmarks parsing in any way. 15 milliseconds in 200 milliseconds is less than 15% of the time. The remaining 85% is just data copying - something you shouldn't even have to do in a solution where the data transfer is better designed.

**Comment by c4milo** [comment]
@alexhultman, again, nobody is stopping you from publishing a blog post with your results.

**Comment by c4milo** [comment]
There is any absolute reason to be upset. I'm sure @felixge knew there will be always a better way. I'm glad you came in to bring a case. That's what open source is all about.

**Comment by ghost** [comment]
I already published my result in my repo. It can be wrapped in a JS addon with no performance loss, as the interface is not about returning a JS object filled with JS strings that holds copies of the data.

My parser just marks where the Utf-8 begins and where it ends, for each field. If you pass it a Node.js buffer with MySql data it would return the same buffer, but with accompanying markers. That way you don't get copies of the data back, but know where each field is located inside the buffer.

One could then, if being super serious about this thing, implement some kind of utf-8 wrapper as opposed to using the built-in v8::String (which does not play well with utf-8).

So basically, the key to performance in this is not about the parser, but instead how you hand off the data to JS. The idea is to keep the data outside of the GC, and outside of JS as much as possible. Any interaction with JavaScript data types would completely kill any parsing advantage and just bring everything down to the same sluggish level.

**Comment by felixge** [comment]
> Just a quick question then. Why did you decide to use JavaScript objects as a means of transporting the data into the script?
> Because we both know that you could easily get 5 - 10x better perf had you chosen any other interface. Basically you disqualify yourself from any opportunity of getting good perf with that decision.

Because I wanted to provide an API that people would find useful.

> Any benchmark that concludes a solution based on Math.pow to read unsigned integers to be "pretty damn fast" is clearly completely flawed.

It isn't. I only applied optimizations when they yielded noticeable performance gain. IIRC `Math.pow()` was not critical as it was only called for parsing the packet header, which was irrelevant to the overall performance.

> The only thing being benchmarked here is "how fast can V8 copy strings, and put them in a JS object".

Besides not knowing the definition of the word "only", you're also completely missing the point.

A significant effort of writing this parser was figuring out how to make the parsing code fast. Once that was done, the costs were indeed dominated by creating v8 strings/objects. But getting there was the whole point of this project.

Anyway, I'll mute this thread now to avoid being baited into your trolling again.
